src/encoding.rs

```diff
@@ -77,6 +77,113 @@ where
     }
 }
 
+/// Additional information passed to every decode/merge function.
+#[derive(Clone, Debug)]
+pub struct DecodeContext {
+    #[cfg(feature = "recursion-limit")]
+    /// How many times we can recurse in the current decode stack before we hit
+    /// the recursion limit.
+    ///
+    /// The recursion limit is defined by `RECURSION_LIMIT` and cannot be
+    /// customized. The recursion limit can be ignored by building the Prost
+    /// crate without the `recursion-limit` feature (which is set by default).
+    recurse_count: u32,
+}
+
+impl Default for DecodeContext {
+    #[cfg(feature = "recursion-limit")]
+    fn default() -> DecodeContext {
+        DecodeContext {
+            recurse_count: crate::RECURSION_LIMIT,
+        }
+    }
+
+    #[cfg(not(feature = "recursion-limit"))]
+    fn default() -> DecodeContext {
+        DecodeContext {}
+    }
+}
+
+impl DecodeContext {
+    #[cfg(feature = "recursion-limit")]
+    /// Call this function before recursively decoding.
+    ///
+    /// This function returns a guard object which will automatically restore
+    /// the recursion counter when it is destroyed by going out of scope.
+    ///
+    /// See the safety note on `RecursionGuard` for important information.
+    pub(crate) fn enter_recursion(&mut self) -> RecursionGuard {
```

Comment:
Yeah I think this can be significantly simplified through passing an owned struct which includes the counter, e.g. `fn recurse(&self) -> Result<Self, DecodeError>` to pass the context, and automatically increment or decrement the counter and check against 0 or max (I don't love the name `recurse`, though, so maybe there's a better option for that).


To enforce that `recurse` is used, the type should opt-out of `Clone` and `Copy`.

Reply:
> To enforce that recurse is used, the type should opt-out of Clone and Copy.

hmm actually that would be problematic for decoding `repeated` values.  Maybe a suitably named method which is effectively a clone but wouldn't be used accidentally would solve that.

Reply:
I think using Clone but opting out of Copy is fine. It doesn't force you to opt-in to recursing, but I don't think you can in general because the context can always be moved. Where there is a loop, it is nice to `clone` rather than implicitly copy I think - there is at least a hint that in some cases there is something else to do.